lib/Parse/ParseDecl.cpp

```diff
@@ -3063,9 +3063,26 @@ ParserStatus Parser::parseNewDeclAttribute(DeclAttributes &Attributes,
     if (!Tok.is(tok::l_paren)) {
       // Normal access control attribute.
       AttrRange = Loc;
-      DuplicateAttribute = Attributes.getAttribute<AccessControlAttr>();
-      if (!DuplicateAttribute)
+      
+      if (!DiscardAttribute) {
         Attributes.add(new (Context) AccessControlAttr(AtLoc, Loc, access));
+        break;
+      }
+      
+      if (access != cast<AccessControlAttr>(DuplicateAttribute)->getAccess()) {
+        // Diagnose if there's already an access control attribute on
+        // this declaration with a different access level.
```

Comment:
This comment summarizes the entire statement, so I think it deserves to be hoisted. Ditto the other case.

Now, some people seem to prefer commenting on statements within their code blocks. I’m not a fan of this style for several reasons:
* It adds to the cognitive load and speed of reading code for me, both because I have to figure out if a comment within a code block pertains to the entire statement or not, and because I’m used to reading the statement before deciding whether to descend into its code block(s).
* If you fold a code block in an IDE, the associated comment disappears.

Reply:
Done! I agree in this case it made a lot of sense to hoist it. And if it's the common practice in the codebase, it would be another great reason to do it.

On the other PR we're having a discussion, I believe there's a great example of outer comments talking about when and why the block of code is run, and inner comments talking about what the block of code does when it runs. To me, it was a great way of documenting the code.

Reply:
> On the other PR we're having a discussion, I believe there's a great example of outer comments talking about when and why the block of code is run, and inner comments talking about what the block of code does when it runs. To me, it was a great way of documenting the code.

This, right?
```c++
// Only allowed to synthesize a throwing super.init() call if the init being
// checked is also throwing.
if (ctor->hasThrows()) {
  // Diagnose on nonthrowing or rethrowing initializer
  if (!fromCtor->hasThrows() || fromCtor->hasPolymorphicEffect(EffectKind::Throws)) {
    fromCtor->diagnose(diag::implicit_throws_super_init);
    return true; // considered an error
  }
}
```

I agree the `hasThrows` behavioral quirk is worth the additional inner clarification. And, given this is a simple code block performing a single action, I really like that you’re at least alluding to the "what" besides talking about the "why" and "when" in the outer comment. On a larger scale, it becomes so much simpler to wrap your head around the logic behind the code with this context arrangement.